### PR TITLE
fix(managers/gradle): allow dot annotation in version.ref

### PR DESCRIPTION
**Pull request by cjtim**
## Changes
`.` `-` `_` in version.ref can be a reference to the version variable

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/gradle/extract/catalog.ts [hunk 1]
```diff
@@ -79,12 +79,14 @@ function extractVersion({
   versionSubContent: string;
 }): VersionExtract {
   if (isVersionPointer(version)) {
+    // This is to support dot annotation version reference
+    const parsedVersion = version.ref.replaceAll('.', '-').replaceAll('_', '-');
```
Comment (Churro) [inline lib/modules/manager/gradle/extract/catalog.ts:83]:
```suggestion
    const parsedVersion = version.ref.replace(/[._]/g, '-');
```

https://jsben.ch/KY0Cv

File: lib/modules/manager/gradle/extract/catalog.ts [hunk 2]
```diff
@@ -79,12 +79,14 @@ function extractVersion({
   versionSubContent: string;
 }): VersionExtract {
   if (isVersionPointer(version)) {
+    // This is to support dot annotation version reference
```
Comment (Churro) [inline lib/modules/manager/gradle/extract/catalog.ts:82]:
```suggestion
```

Not complex enough to need additional explanation.

File: lib/modules/manager/gradle/extract/catalog.ts [hunk 3]
```diff
@@ -205,7 +207,7 @@ function extractDependency({
     };
   }
   const versionRef = isVersionPointer(descriptor.version)
-    ? descriptor.version.ref
+    ? descriptor.version.ref.replaceAll('.', '-').replaceAll('_', '-')
```
Comment (Churro) [inline lib/modules/manager/gradle/extract/catalog.ts:210]:
```suggestion
    ? descriptor.version.ref.replace(/[._]/g, '-')
```
https://jsben.ch/KY0Cv

**Review by Churro — CHANGES_REQUESTED** [review]
- Documentation update needed? Please tick accordingly
- When proposing such changes, please provide evidence like a reference to Gradle docs / code or a minimum reproduction repo that demonstrates practical viability
- I think the change should also be done here: https://github.com/renovatebot/renovate/blob/dc1d1e85db8ca4d8640af438ba9834da0ca276f9/lib/modules/manager/gradle/extract/catalog.ts#L287

File: lib/modules/manager/gradle/extract/catalog.ts (hunk 3 above)
Comment (cjtim) [reply lib/modules/manager/gradle/extract/catalog.ts:210]: @Churro 
do i need to use `regEx(/[._]/g`) library with this?

Comment (Churro) [reply lib/modules/manager/gradle/extract/catalog.ts]: make sense 👍

**Review by viceice — COMMENTED** [review]
failed code coverage

**Comment by cjtim** [comment]
> * Documentation update needed? Please tick accordingly
> * When proposing such changes, please provide evidence like a reference to Gradle docs / code or a minimum reproduction repo that demonstrates practical viability
> * I think the change should also be done here: https://github.com/renovatebot/renovate/blob/dc1d1e85db8ca4d8640af438ba9834da0ca276f9/lib/modules/manager/gradle/extract/catalog.ts#L287

@Churro 
I got example repo and ran with this PR's code https://github.com/cjtim/gradle-example 

@viceice CodeCov not highlight lines that missing test case not sure what to do

**Review by viceice — COMMENTED** [review]
Strange indirect coverage change:
https://app.codecov.io/gh/renovatebot/renovate/pull/21820/indirect-changes

File: lib/modules/manager/gradle/extract/catalog.ts [hunk 4]
```diff
@@ -205,7 +206,7 @@ function extractDependency({
     };
   }
   const versionRef = isVersionPointer(descriptor.version)
-    ? descriptor.version.ref
+    ? descriptor.version.ref.replace(regEx(/[._]/g), '-')
```
[the remark this answers is not in the document]
Comment (cjtim) [reply lib/modules/manager/gradle/extract/catalog.ts:209]: no, it cannot. toml parser will throw an error if toml is invalid

```toml
[abc]
foo = { version.ref = }
```

**Comment by cjtim** [comment]
@Churro I update as you suggested. Please review it again thanks! :)

File: lib/modules/manager/gradle/extract/catalog.ts [hunk 5]
```diff
@@ -55,6 +55,10 @@ function isVersionPointer(
   return hasKey('ref', obj);
 }
 
+function normalizeVersionPointer(versionPointer: string): string {
+  return versionPointer?.replace(regEx(/[._]/g), '-');
```
Comment (Churro) [inline lib/modules/manager/gradle/extract/catalog.ts:59]:
```suggestion
  return versionPointer.replace(regEx(/[._]/g), '-');
```

`versionPointer` can't be optional here

**Review by Churro — CHANGES_REQUESTED** [review]

File: lib/modules/manager/gradle/extract/catalog.ts (hunk 5 above)
Comment (cjtim) [reply lib/modules/manager/gradle/extract/catalog.ts]: fixed

**Review by Churro — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**